scylla/src/frame/mod.rs

```diff
@@ -17,21 +17,69 @@ mod cql_types_test;
 use crate::frame::frame_errors::FrameError;
 use crate::transport::Compression;
 use bytes::{Buf, BufMut, Bytes};
-use tokio::io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt};
+use tokio::io::{AsyncRead, AsyncReadExt};
 use uuid::Uuid;
 
 use std::convert::TryFrom;
 
 use compress::lz4;
-use request::RequestOpcode;
+use request::Request;
 use response::ResponseOpcode;
 
+const HEADER_SIZE: usize = 9;
+
 // Frame flags
 pub const FLAG_COMPRESSION: u8 = 0x01;
 pub const FLAG_TRACING: u8 = 0x02;
 pub const FLAG_CUSTOM_PAYLOAD: u8 = 0x04;
 pub const FLAG_WARNING: u8 = 0x08;
 
+pub struct SerializedRequest {
+    data: Vec<u8>,
+}
+
+// TODO: Should we support serialization in other versions of the protocol than V4?
+impl SerializedRequest {
+    pub fn make<R: Request>(
+        req: &R,
+        compression: Option<Compression>,
+        tracing: bool,
+    ) -> Result<SerializedRequest, FrameError> {
+        let mut flags = 0;
+        let mut data = vec![0; HEADER_SIZE];
+
+        if let Some(compression) = compression {
+            flags |= FLAG_COMPRESSION;
+            let body = req.to_bytes()?;
+            compress_append(&body, compression, &mut data)?;
+        } else {
+            req.serialize(&mut data)?;
+        }
+
+        if tracing {
+            flags |= FLAG_TRACING;
+        }
+
+        data[0] = 4; // We only support version 4 for now
+        data[1] = flags;
+        // Leave space for the stream number
+        data[4] = R::OPCODE as u8;
+
+        let req_size = (data.len() - HEADER_SIZE) as u32;
+        data[5..9].copy_from_slice(&req_size.to_be_bytes());
+
+        Ok(Self { data })
+    }
+
+    pub fn set_stream(&mut self, stream: i16) {
```

Comment:
Why can't we pass stream ID directly to `make()` function?

Reply:
I think it's because a new stream id is allocated later in the flow and setting the stream id later makes the change much less intrusive, but I'll wait for @piodul's clarification